boundary: account for TEE space with RAM top addr

tbensonatl opened the pull request:
Account for TEE size in IMX8M version of board_get_usable_ram_top, which is used to find the U-Boot relocation address in common/board_f.c. This logic is similar to the address and RAM size adjustments in arch/arm/mach-imx/imx8m/soc.c.

gibsson commented:
Hi,
Thanks for the contribution. Can you add your Signed-off-by and modify the title to have "boundary:" to mention this is part of the board/boundary only.
Can you confirm that rom_pointer is always set, no matter the platform?
Thanks

tbensonatl commented:
Hi,

I updated the commit message and added a Signed-Off-By. The function modified in this patch is in a define guard for i.MX8M boards (#if defined(CONFIG_IMX8M)), so the change only applies to boards for which that config option is set. The startup arguments (via registers x1, x2, etc.) are saved to rom_pointer for the i.MX8M in arch/arm/mach-imx/imx8m/lowlevel_init.S. These registers are populated by the ARM Trusted Firmware (e.g. at https://github.com/boundarydevices/imx-atf/blob/boundary-lf-5.10.72-2.2.0/plat/imx/imx8m/imx8mm/imx8mm_bl31_setup.c#L172), but if another boot flow uses a different convention then the rom_pointer values may not be correctly populated. I confirmed that my Nitrogen i.MX8M Mini board boots correctly whether or not I include OP-TEE (i.e. rom_pointer[1] is 0), but I have not been able to test on any other boards.

Regards

gibsson commented:
Hi,
Thanks for the quick respin and explanation, it looks good.
@tkisky any objections?
Regards

tkisky commented:
On 8/15/2022 8:19 AM, Gary Bisson wrote:
> Hi,
> Thanks for the quick respin and explanation, it looks good.
> @tkisky <https://github.com/tkisky> any objections?
> Regards
> 
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/boundarydevices/u-boot/pull/16#issuecomment-1215144717>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANA3P2WWMXGUY7MXJSFUPLVZJNYRANCNFSM55ZFR3MQ>.
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> 
I think this looks better


if ((rom_pointer[0] > CONFIG_SYS_SDRAM_BASE) &&
    (rom_pointer[0] < (CONFIG_SYS_SDRAM_BASE + MEM_SIZE)) &&
    ((rom_pointer[0] + rom_pointer[1]) >= (MEM_SIZE + CONFIG_SYS_SDRAM_BASE)))
	return rom_pointer[0];
else
	return CONFIG_SYS_SDRAM_BASE + MEM_SIZE;



I don't think optee is always at the end of memory.

tbensonatl commented:
One note is that `dram_init()` in `arch/arm/mach-imx/imx8m/soc.c` will set `gd->ram_size` to `sdram_size - rom_pointer[1]`. That may still cause issues in the case of OP-TEE not being at the end of memory if other code assumes that `gd->ram_size` contiguous bytes are available starting at the base address. I agree that the sanity checks on `rom_pointer[0]` are a good idea, but it seems that the `(rom_pointer[0] + rom_pointer[1]) >= (MEM_SIZE + CONFIG_SYS_SDRAM_BASE))` clause also assumes that the TEE occupies the top of memory. Or probably that was meant to be a <=?
